lib/mongoid/persistable.rb

```diff
@@ -44,24 +44,80 @@ module Persistable
+    # that would happen inside the block. This method supports nesting further
+    # calls to atomically, which will behave according to the options described
+    # below.
+    #
+    # An option join_context can be given which, when true, will merge the
+    # operations declared by the given block with the atomically block wrapping
+    # the current invocation, if one exists. If this block or any other block
+    # sharing the same context raises before persisting, then all the operations
+    # of that context will not be persisted, and will also be reset in-memory.
+    #
+    # When join_context is false, the given block of operations will be
+    # persisted independently of other contexts. Failures in other contexts will
+    # not affect this one, so long as this block was able to run and persist
+    # changes.
+    #
+    # The default value of join_context is set by the global configuration
+    # option atomically_join_context_default, whose own default is false.
     #
     # @example Execute the operations atomically.
     #   document.atomically do
     #     document.set(name: "Tool").inc(likes: 10)
     #   end
     #
+    # @example Execute some inner operations atomically, but independently from the outer operations.
+    #
+    #   document.atomically do
+    #     document.inc likes: 10
+    #     document.atomically join_context: false do
+    #       # The following is persisted to the database independently.
+    #       document.unset :origin
+    #     end
+    #     document.atomically join_context: true do
+    #       # The following is persisted along with the other outer operations.
+    #       document.inc member_count: 3
+    #     end
+    #     document.set name: "Tool"
+    #   end
+    #
+    # @param [ Hash ] options The options to pass to atomically.
+    #
+    # @option options [ true, false ] :join_context Join the context (i.e. merge
+    #   declared atomic operations) of the atomically block wrapping this one,
+    #   if one exists.
+    #
     # @return [ true, false ] If the operation succeeded.
     #
     # @since 4.0.0
-    def atomically
-      begin
-        @atomic_updates_to_execute = @atomic_updates_to_execute || {}
-        yield(self) if block_given?
-        persist_atomic_operations(@atomic_updates_to_execute)
-        true
-      ensure
-        @atomic_updates_to_execute = nil
+    def atomically(join_context: Mongoid.atomically_join_context_default)
+      call_depth = @atomically_depth ||= 0
+      has_own_context = call_depth.zero? || !join_context
+      @atomic_updates_to_execute_stack ||= []
+      push_atomically_context if has_own_context
+
+      if block_given?
+        @atomically_depth += 1
+        yield(self)
+        @atomically_depth -= 1
+      end
+
+      if has_own_context
+        persist_atomic_operations @atomically_context
+        remove_atomically_context_changes
+      end
+
+      true
+    rescue StandardError => e
```

Comment:
This should be `Exception` I think. `StandardError` is more for rescues that handle exceptions, but when we clean up we should do so for all errors.